File: module2/class1/exemplo18.cs
/*
================================================================================
ATIVIDADE PRÁTICA 18 - JSON SERIALIZATION PERFORMANCE (C#)
================================================================================

OBJETIVO:
- Demonstrar overhead de JSON serialization repeated em loops
- Usar CPU profiler para identificar tempo gasto em serialization
- Otimizar usando JsonSerializer com options caching
- Comparar Newtonsoft.Json vs System.Text.Json performance

PROBLEMA:
- Repeated serialization/deserialization é custosa
- Reflection overhead em serializers
- CPU Profiler mostrará tempo gasto em JSON operations

SOLUÇÃO:
- Cache serializer options e settings
- Use System.Text.Json para better performance
- Serialize em batches quando possível

================================================================================
*/

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.Json;
using Newtonsoft.Json;

class DataRecord {
    public int Id { get; set; }
    public string Name { get; set; }
    public double Value { get; set; }
    public DateTime Timestamp { get; set; }
    public bool IsActive { get; set; }
    public List<string> Tags { get; set; }
    
    public static DataRecord CreateSample(int id) {
        return new DataRecord {
            Id = id,
            Name = $"Record_{id:D6}",
            Value = id * 3.14159,
            Timestamp = DateTime.Now.AddMinutes(-id),
            IsActive = id % 2 == 0,
            Tags = new List<string> { $"tag_{id % 3}", $"category_{id % 5}" }
        };
    }
}

class Program {
    static void DemonstrateInefficiientSerialization() {
        Console.WriteLine("Starting inefficient JSON serialization demonstration...");
        Console.WriteLine("Monitor CPU profiler - should see time spent in JSON operations");
        
        const int ITERATIONS = 10000;
        var records = new List<DataRecord>();
        var jsonResults = new List<string>();
        
        // Create test data
        for (int i = 0; i < ITERATIONS; i++) {
            records.Add(DataRecord.CreateSample(i));
        }
        
        var sw = Stopwatch.StartNew();
        
        for (int i = 0; i < ITERATIONS; i++) {
            // PERFORMANCE ISSUE: Creating new JsonSerializerSettings every time
            var settings = new JsonSerializerSettings {
                DateFormatHandling = DateFormatHandling.IsoDateFormat,
                NullValueHandling = NullValueHandling.Ignore,
                Formatting = Formatting.None
            };
            
            // PERFORMANCE ISSUE: Newtonsoft.Json with repeated settings creation
            string json = JsonConvert.SerializeObject(records[i], settings);
            jsonResults.Add(json);
            
            // PERFORMANCE ISSUE: Immediate deserialization with new settings
            var deserializedSettings = new JsonSerializerSettings {
                DateFormatHandling = DateFormatHandling.IsoDateFormat,
                NullValueHandling = NullValueHandling.Ignore
            };
            
            DataRecord deserialized = JsonConvert.DeserializeObject<DataRecord>(json, deserializedSettings);
            
            if (i % 1000 == 0) {
                Console.WriteLine($"Serialized {i}/{ITERATIONS} objects...");
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Inefficient JSON serialization completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total objects serialized: {jsonResults.Count}");
        Console.WriteLine($"Average JSON size: {jsonResults[0].Length} chars");
    }
    
    static void Main() {
        Console.WriteLine("Starting JSON serialization performance demonstration...");
        Console.WriteLine("Task: Serializing objects with repeated settings creation");
        Console.WriteLine("Monitor CPU Usage Tool for JSON serialization overhead");
        Console.WriteLine();
        
        DemonstrateInefficiientSerialization();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check CPU profiler for:");
        Console.WriteLine("- Time spent in JsonConvert.SerializeObject");
        Console.WriteLine("- Reflection overhead in serialization");
        Console.WriteLine("- Repeated settings object creation");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR EFFICIENT SERIALIZATION)
================================================================================

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.Json;

class DataRecord {
    public int Id { get; set; }
    public string Name { get; set; }
    public double Value { get; set; }
    public DateTime Timestamp { get; set; }
    public bool IsActive { get; set; }
    public List<string> Tags { get; set; }
    
    public static DataRecord CreateSample(int id) {
        return new DataRecord {
            Id = id,
            Name = $"Record_{id:D6}",
            Value = id * 3.14159,
            Timestamp = DateTime.Now.AddMinutes(-id),
            IsActive = id % 2 == 0,
            Tags = new List<string> { $"tag_{id % 3}", $"category_{id % 5}" }
        };
    }
}

class Program {
    // CORREÇÃO: Cache JsonSerializerOptions to avoid repeated creation
    private static readonly JsonSerializerOptions CachedOptions = new JsonSerializerOptions {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = false,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
    };
    
    static void DemonstrateEfficientSerialization() {
        Console.WriteLine("Starting efficient JSON serialization demonstration...");
        Console.WriteLine("Monitor CPU profiler - should see reduced JSON overhead");
        
        const int ITERATIONS = 10000;
        var records = new List<DataRecord>();
        var jsonResults = new List<string>();
        
        // Create test data
        for (int i = 0; i < ITERATIONS; i++) {
            records.Add(DataRecord.CreateSample(i));
        }
        
        var sw = Stopwatch.StartNew();
        
        for (int i = 0; i < ITERATIONS; i++) {
            // CORREÇÃO: Use cached options with System.Text.Json (faster than Newtonsoft)
            string json = System.Text.Json.JsonSerializer.Serialize(records[i], CachedOptions);
            jsonResults.Add(json);
            
            // CORREÇÃO: Reuse same cached options for deserialization
            DataRecord deserialized = System.Text.Json.JsonSerializer.Deserialize<DataRecord>(json, CachedOptions);
            
            if (i % 1000 == 0) {
                Console.WriteLine($"Efficiently serialized {i}/{ITERATIONS} objects...");
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Efficient JSON serialization completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total objects serialized: {jsonResults.Count}");
        Console.WriteLine($"Using System.Text.Json with cached options");
    }
    
    static void DemonstrateBatchSerialization() {
        Console.WriteLine("Starting batch serialization demonstration...");
        
        const int ITERATIONS = 10000;
        const int BATCH_SIZE = 1000;
        
        var records = new List<DataRecord>();
        var batchResults = new List<string>();
        
        // Create test data
        for (int i = 0; i < ITERATIONS; i++) {
            records.Add(DataRecord.CreateSample(i));
        }
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: Serialize in batches for better efficiency
        for (int i = 0; i < records.Count; i += BATCH_SIZE) {
            var batch = records.GetRange(i, Math.Min(BATCH_SIZE, records.Count - i));
            
            // Serialize entire batch at once - much more efficient
            string batchJson = System.Text.Json.JsonSerializer.Serialize(batch, CachedOptions);
            batchResults.Add(batchJson);
            
            if (i % 2000 == 0) {
                Console.WriteLine($"Batch serialized {i}/{records.Count} objects...");
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Batch serialization completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total batches: {batchResults.Count}");
        Console.WriteLine($"Objects per batch: {BATCH_SIZE}");
    }
    
    static void DemonstrateMemoryEfficientSerialization() {
        Console.WriteLine("Starting memory-efficient serialization...");
        
        const int ITERATIONS = 10000;
        var records = new List<DataRecord>();
        
        for (int i = 0; i < ITERATIONS; i++) {
            records.Add(DataRecord.CreateSample(i));
        }
        
        var sw = Stopwatch.StartNew();
        long totalBytes = 0;
        
        // CORREÇÃO: Use Utf8JsonWriter for maximum performance
        for (int i = 0; i < ITERATIONS; i++) {
            using var buffer = new System.Buffers.ArrayBufferWriter<byte>();
            using var writer = new Utf8JsonWriter(buffer);
            
            System.Text.Json.JsonSerializer.Serialize(writer, records[i], CachedOptions);
            totalBytes += buffer.WrittenCount;
        }
        
        sw.Stop();
        
        Console.WriteLine($"Memory-efficient serialization completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Total bytes written: {totalBytes:N0}");
        Console.WriteLine($"Using Utf8JsonWriter for maximum performance");
    }
    
    static void Main() {
        Console.WriteLine("Starting optimized JSON serialization demonstration...");
        Console.WriteLine("Task: Efficient JSON serialization strategies");
        Console.WriteLine("Monitor CPU Usage Tool for improved performance");
        Console.WriteLine();
        
        DemonstrateEfficientSerialization();
        Console.WriteLine();
        DemonstrateBatchSerialization();
        Console.WriteLine();
        DemonstrateMemoryEfficientSerialization();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- System.Text.Json is faster than Newtonsoft.Json");
        Console.WriteLine("- Cached options eliminate repeated setup overhead");
        Console.WriteLine("- Batch serialization reduces per-object overhead");
        Console.WriteLine("- Utf8JsonWriter provides maximum performance");
        Console.WriteLine("- Significantly reduced CPU usage and memory allocations");
    }
}

================================================================================
*/

File: module2/class1/exemplo18.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 18 - STRING FORMATTING PERFORMANCE (C++)
================================================================================

OBJETIVO:
- Demonstrar overhead de string formatting repeated em loops
- Usar CPU profiler para identificar tempo gasto em sprintf/stringstream
- Otimizar usando buffer reuse e efficient formatting
- Comparar diferentes approaches de string formatting

PROBLEMA:
- Repeated sprintf/stringstream operations são custosas
- Memory allocations em string operations
- CPU Profiler mostrará tempo gasto em formatting functions

SOLUÇÃO:
- Reuse buffers quando possível
- Use more efficient formatting methods
- Pre-allocate string capacity

================================================================================
*/

#include <iostream>
#include <string>
#include <sstream>
#include <chrono>
#include <vector>
#include <iomanip>
using namespace std;

void demonstrateInefficiientStringFormatting() {
    cout << "Starting inefficient string formatting demonstration..." << endl;
    cout << "Monitor CPU profiler - should see time spent in string operations" << endl;
    
    const int ITERATIONS = 50000;
    vector<string> formattedStrings;
    formattedStrings.reserve(ITERATIONS);
    
    auto start = chrono::high_resolution_clock::now();
    
    for (int i = 0; i < ITERATIONS; i++) {
        // PERFORMANCE ISSUE: Creating new stringstream every iteration
        stringstream ss;
        ss << "Record #" << setfill('0') << setw(6) << i 
           << " - Value: " << fixed << setprecision(2) << (i * 3.14159) 
           << " - Status: " << (i % 2 == 0 ? "ACTIVE" : "INACTIVE");
        
        string formatted = ss.str(); // String allocation and copy
        formattedStrings.push_back(formatted);
        
        // PERFORMANCE ISSUE: Another string operation
        string upperCased;
        for (char c : formatted) {
            upperCased += toupper(c); // Character by character concatenation
        }
        
        if (i % 5000 == 0) {
            cout << "Formatted " << i << "/" << ITERATIONS << " strings..." << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Inefficient string formatting completed in: " << duration.count() << " ms" << endl;
    cout << "Total strings created: " << formattedStrings.size() << endl;
    cout << "Example result: " << formattedStrings[100] << endl;
}

int main() {
    cout << "Starting string formatting performance demonstration..." << endl;
    cout << "Task: Formatting many strings with repeated stringstream creation" << endl;
    cout << "Monitor CPU Usage Tool for string formatting overhead" << endl << endl;
    
    demonstrateInefficiientStringFormatting();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check CPU profiler for:" << endl;
    cout << "- Time spent in stringstream operations" << endl;
    cout << "- String allocation and deallocation overhead" << endl;
    cout << "- Character-by-character string building" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR EFFICIENT STRING FORMATTING)
================================================================================

#include <iostream>
#include <string>
#include <sstream>
#include <chrono>
#include <vector>
#include <iomanip>
#include <cstdio>
using namespace std;

void demonstrateEfficientStringFormatting() {
    cout << "Starting efficient string formatting demonstration..." << endl;
    cout << "Monitor CPU profiler - should see reduced string operation overhead" << endl;
    
    const int ITERATIONS = 50000;
    vector<string> formattedStrings;
    formattedStrings.reserve(ITERATIONS);
    
    // CORREÇÃO: Reuse stringstream instead of creating new ones
    stringstream ss;
    
    // CORREÇÃO: Pre-allocate buffer for sprintf
    char buffer[256];
    
    auto start = chrono::high_resolution_clock::now();
    
    for (int i = 0; i < ITERATIONS; i++) {
        // CORREÇÃO: Clear and reuse existing stringstream
        ss.str("");
        ss.clear();
        
        // Option 1: Efficient stringstream reuse
        ss << "Record #" << setfill('0') << setw(6) << i 
           << " - Value: " << fixed << setprecision(2) << (i * 3.14159) 
           << " - Status: " << (i % 2 == 0 ? "ACTIVE" : "INACTIVE");
        
        string formatted = ss.str();
        
        // CORREÇÃO: Reserve capacity for string operations
        string upperCased;
        upperCased.reserve(formatted.length());
        
        // CORREÇÃO: More efficient character transformation
        for (char c : formatted) {
            upperCased.push_back(toupper(c));
        }
        
        formattedStrings.push_back(move(formatted));
        
        if (i % 5000 == 0) {
            cout << "Efficiently formatted " << i << "/" << ITERATIONS << " strings..." << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Efficient string formatting completed in: " << duration.count() << " ms" << endl;
    cout << "Total strings created: " << formattedStrings.size() << endl;
    cout << "Example result: " << formattedStrings[100] << endl;
}

void demonstrateSprintfApproach() {
    cout << "Starting sprintf approach demonstration..." << endl;
    
    const int ITERATIONS = 50000;
    vector<string> formattedStrings;
    formattedStrings.reserve(ITERATIONS);
    
    // CORREÇÃO: Use sprintf for maximum formatting performance
    char buffer[256];
    
    auto start = chrono::high_resolution_clock::now();
    
    for (int i = 0; i < ITERATIONS; i++) {
        // CORREÇÃO: sprintf is often faster than stringstream for simple formatting
        snprintf(buffer, sizeof(buffer), 
                "Record #%06d - Value: %.2f - Status: %s", 
                i, i * 3.14159, (i % 2 == 0 ? "ACTIVE" : "INACTIVE"));
        
        string formatted(buffer);
        
        // Convert to uppercase efficiently
        for (char& c : formatted) {
            c = toupper(c);
        }
        
        formattedStrings.push_back(move(formatted));
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "sprintf approach completed in: " << duration.count() << " ms" << endl;
    cout << "Example result: " << formattedStrings[100] << endl;
}

int main() {
    cout << "Starting optimized string formatting demonstration..." << endl;
    cout << "Task: Efficient string formatting with buffer reuse" << endl;
    cout << "Monitor CPU Usage Tool for improved performance" << endl << endl;
    
    demonstrateEfficientStringFormatting();
    cout << endl;
    demonstrateSprintfApproach();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- Stringstream reuse eliminates repeated construction" << endl;
    cout << "- String capacity reservation reduces reallocations" << endl;
    cout << "- sprintf can be faster for simple formatting" << endl;
    cout << "- Move semantics reduce copying" << endl;
    cout << "- In-place character transformation is more efficient" << endl;
    
    return 0;
}

================================================================================
*/
